### ISSUE TITLE
scattermoe + FSDP2 LoRA: saved adapter contains only 1/world_size of each lora_B (rest exactly 0.0), regardless of state_dict_type

### ISSUE THREAD

**Issue by trevorgordon981** [axolotl-ai-cloud/axolotl#3772]
### Summary

When training a LoRA on the fused MoE experts via the **scattermoe** integration under **FSDP2** (`glm_moe_dsa` / NVFP4 recipe), the saved `adapter_model.safetensors` contains only **1/world_size** of each `lora_B` matrix — only rank-0's shard is nonzero; the remaining `(world_size-1)/world_size` output rows are **exactly 0.0**. The `lora_A` matrices are fully dense/correct. The resulting adapter is silently unusable (on an 8-GPU run, 7/8 of the learned delta is missing, and the "up" half of `gate_up` gets zero adaptation entirely).

Because the file has valid shapes and full size, integrity checks (size / sha256 / safetensors header) all pass — the only way to catch it is inspecting the nonzero fraction of the weights.

### Setup

- Model: GLM-5.2, NVFP4 base, `glm_moe_dsa` recipe
- LoRA on the routed experts: `lora_target_parameters: [experts.gate_up_proj, experts.down_proj]`, r=32
- FSDP2, `world_size=8`, 8×B200
- scattermoe expert-LoRA path (`axolotl.integrations.kernels.libs.scattermoe_lora.*`, `nvfp4_*`)

### Symptom (numerically, per MoE layer)

For every layer, both `...mlp.experts.lora_B.weight` and `...mlp.experts.base_layer.lora_B.weight`:

- **exactly 12.5% (= 1/8) of output rows nonzero**, e.g. rows `0:512` of 4096 for the gate_up `lora_B`, rows `0:768` of 6144 for the down `lora_B`; every other row is exactly `0.0`.
- `lora_A`: 100% dense (correct).
- Identical pattern across all MoE layers and all 256 experts.

### Reproduced across three save/consolidation paths

- `state_dict_type: SHARDED_STATE_DICT` (consolidated) → 12.5%
- `state_dict_type: FULL_STATE_DICT` → **12.5%** (so it is not a state-dict-type / consolidation setting)
- `torch.distributed.checkpoint` offline re-consolidation of the sharded checkpoint's `.distcp` shards → 12.5%

The fact that the nonzero fraction is *exactly* `1/world_size` strongly suggests the scattermoe LoRA parameters' state-dict path returns only the local (rank-0) shard with no cross-rank all-gather at save time — or, alternatively, that gradient only reaches rank-0's shard of `lora_B`.

### Impact

Training completes cleanly (healthy loss, generalizing eval), a full-size adapter is written, and everything *looks* fine — but the adapter is ~7/8 empty. This is a silent-corruption footgun for anyone doing scattermoe/FSDP2 LoRA on MoE experts.

### Question / ask

- Is this a known issue, and is there a correct FSDP2 wrapping policy to keep the small scattermoe LoRA parameters **replicated** (or a correct all-gather at save) so the full `lora_B` is materialized?
- Happy to provide a minimal repro, the exact config, or test a patch.

**Comment by NanoCode012** [comment]
Hello,

> state_dict_type: SHARDED_STATE_DICT (consolidated) → 12.5%

In this case, is there no merge script to combine back to full? Is it just the full state dict that has an issue based on what you're saying?

**Comment by winglian** [comment]
could you provide a repro or exact config that you used? did you happen to use EP? thanks for reporting this!

**Comment by trevorgordon981** [comment]
Thanks both for taking a look!

**@winglian — no EP.** The run that produced the 12.5% adapter was plain FSDP2 + scattermoe, no ExpertParallel — the only plugins were `CutCrossEntropyPlugin` + `KernelsPlugin`; there was no `ExpertParallelPlugin` and no `expert_parallel_*` keys. (I had an earlier EP-based config, but it was abandoned before it ever trained because DeepEP wouldn't build — that's the separate #3766.) So this reproduces on the plain FSDP2 scattermoe-LoRA path with EP off. Exact config below.

**@NanoCode012 — it's not a merge/consolidation issue; there's nothing to merge back.** All three save paths give exactly 1/world_size (12.5%):

1. `SHARDED_STATE_DICT` → consolidated adapter: 12.5% nonzero rows in every `lora_B`
2. Retest with `state_dict_type: FULL_STATE_DICT`: the full state dict itself is 7/8 zeros — same 12.5%
3. Offline `torch.distributed.checkpoint` re-consolidation of the raw `.distcp` shards: also 12.5%

So the missing rows aren't sitting in some shard that a merge script failed to gather — they're never written. That points to either (i) the scattermoe LoRA parameters' state-dict path returning only the local rank-0 shard with no cross-rank all-gather at save, or (ii) gradient only ever reaching rank-0's shard of `lora_B`. Notably `lora_A` is fully dense/correct, which is why it looks like a `lora_B`-specific gather-at-save or grad-routing issue rather than a general save bug.

**Exact symptom, restated:** on 8x GPUs (world_size=8), per MoE layer: `gate_up` `lora_B` is `[4096, 8192]` with exactly 512 nonzero rows (rows 0:512); `down` `lora_B` is `[6144, 8192]` with exactly 768 nonzero rows (rows 0:768); every other row is exactly 0.0. Identical across all MoE layers and all 256 experts. Inspection was a simple per-row L2-norm nonzero count on the saved adapter's `...mlp.experts.lora_B.weight` / `...experts.base_layer.lora_B.weight` tensors.

**Exact config used** (GLM-5.2 `glm_moe_dsa` MoE LoRA on an NVFP4/modelopt checkpoint, 8x B200 sm100, 1 epoch):

```yaml
# GLM-5.2 (glm_moe_dsa) MoE LoRA on an NVFP4 (modelopt) checkpoint. 8x B200 (sm100), 1 epoch.
base_model: /workspace/glm-nvfp4
plugins:
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
  - axolotl.integrations.kernels.KernelsPlugin
use_kernels: true
use_scattermoe: true
experts_implementation: scattermoe
dsv4_fp4_grouped_mode: nvfp4
moe_grouped_backend: deepgemm
use_glm_dsa_kernels: true
strict: false
attn_implementation: sdpa
chat_template: tokenizer_default
datasets:
  - path: /workspace/train.jsonl
    type: chat_template
    field_messages: messages
val_set_size: 0.02
sequence_len: 4096
sample_packing: true
adapter: lora
lora_r: 32
lora_alpha: 32
lora_dropout: 0
lora_qkv_kernel: false
lora_o_kernel: false
lora_mlp_kernel: false
lora_target_modules: 'model.layers.[\d]+.(self_attn.(q_a_proj|q_b_proj|kv_a_proj_with_mqa|kv_b_proj|o_proj)|mlp.(shared_experts.)?(gate_proj|up_proj|down_proj))'
lora_target_parameters:
  - experts.gate_up_proj
  - experts.down_proj
gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 1
optimizer: adamw_torch_8bit
lr_scheduler: cosine
learning_rate: 0.0002
bf16: auto
tf32: true
gradient_checkpointing: true
activation_offloading: true
fsdp_version: 2
fsdp_config:
  offload_params: false
  cpu_ram_efficient_loading: true
  state_dict_type: SHARDED_STATE_DICT
  auto_wrap_policy: TRANSFORMER_BASED_WRAP
  transformer_layer_cls_to_wrap: GlmMoeDsaDecoderLayer
  reshard_after_forward: true
```

(The FULL_STATE_DICT retest used this same config with only `state_dict_type` changed — same 12.5% result.)

One caveat: the training GPU box has since been torn down, so the above is the exact config rather than a live re-run. Happy to spin up a minimal repro or test a patch if you can point at a candidate fix.

**Comment by winglian** [comment]
@trevorgordon981 was able to reproduce locally on a smaller moe. https://github.com/axolotl-ai-cloud/axolotl/pull/3779/changes should fix it

**Comment by trevorgordon981** [comment]
Great, thanks for the fast repro and fix.

Reading #3779, it lines up with hypothesis (i) from above — save-path gather, not grad-routing. The adapter-gather path (`_save_fsdp2_quantized_lora_adapter`) was gated on the trainer's `is_fsdp_enabled` flag and on quant detection via `_local_tensor` only; in my config that flag wasn't set (FSDP2 came in through `fsdp_config` rather than HF/accelerate detection) and the NVFP4 subclass sits in `param.data`, so both gates missed and it silently fell back to the DCP sharded save that only writes rank-0's shard — hence exactly 1/world_size (12.5%) nonzero rows in every `lora_B`, with `lora_A` dense. Routing through `DTensor.full_tensor` and widening detection to `.data` covers both.

Good news is that means the weights themselves were trained correctly across all shards; only the checkpoint export dropped 7/8 of `lora_B`.

My B200 box is torn down so I can't rerun the exact GLM-5.2 config right now, but I'll pull the branch and verify per-row nonzero counts on the saved adapter next time I spin one up (or against your smaller-MoE repro) and report back. Thanks again.
